src/main/java/net/neoforged/neoforge/transfer/resources/IResourceStack.java

```diff
@@ -0,0 +1,164 @@
+ *
+ * @param <T> resource type
+ */
+public interface IResourceStack<T extends IResource> {
+    /**
+     * Creates a codec with the resource being a field in the object.
+     *
+     * <pre>{@code
+     * {
+     *     "resource": {
+     *         "id": "minecraft:water",
+     *         "components": { ... }
+     *     },
+     *     "amount": 1000
+     * }
+     * }</pre>
+     *
+     * @param resourceCodec a codec for the resource
+     * @param <R>           the resource type
+     * @return a codec for a resource stack
+     */
+    static <R extends IResource, S extends IResourceStack<R>> Codec<S> codec(Codec<R> resourceCodec, IStackFactory<R, S> factory) {
+        return RecordCodecBuilder.create(instance -> instance.group(
+                resourceCodec.fieldOf("resource").forGetter(IResourceStack<R>::resource),
+                NeoForgeExtraCodecs.optionalFieldAlwaysWrite(ExtraCodecs.NON_NEGATIVE_INT, "amount", 1)
+                        .forGetter(IResourceStack<R>::amount))
+                .apply(instance, factory::create));
+    }
+
+    /**
+     * Creates a codec where the fields for the resource are at the same level as the amount
+     *
+     * <pre>{@code
+     * {
+     *    "id": "minecraft:water",
+     *    "components": { ... },
+     *    "amount": 1000
+     * }
+     * }</pre>
+     *
+     * @param resourceCodec Backing resource codec
+     * @param factory       Constructor of IResourceStack implementation
+     * @param <R>           The resource type
+     * @return Codec for the specified IResourceStack implementer
+     * @deprecated Use {@link #codec(Codec, IStackFactory)} instead. This sadly can't be used safely with the map when dealing with "empty" resources
+     */
+    //TODO currently doesn't handle empty resources. Due to how the id is missing with . We may want to have the id be "id": "empty" when not present.
+    @Deprecated(forRemoval = true, since = "now")
+    static <R extends IResource, S extends IResourceStack<R>> Codec<S> flatCodec(Codec<R> resourceCodec, IStackFactory<R, S> factory) {
+        return RecordCodecBuilder.create(instance -> instance.group(
+                MapCodec.assumeMapUnsafe(resourceCodec).forGetter(IResourceStack<R>::resource),
+                NeoForgeExtraCodecs.optionalFieldAlwaysWrite(ExtraCodecs.NON_NEGATIVE_INT, "amount", 1)
+                        .forGetter(IResourceStack<R>::amount))
+                .apply(instance, factory::create));
+    }
+
+    /**
+     * Creates a standard stream codec for a IResourceStack implementer of the specified resource type.
+     */
+    static <B extends FriendlyByteBuf, R extends IResource, S extends IResourceStack<R>> StreamCodec<B, S> streamCodec(StreamCodec<? super B, R> resourceCodec, IStackFactory<R, S> factory) {
+        return StreamCodec.composite(
+                resourceCodec, IResourceStack::resource,
+                ByteBufCodecs.VAR_INT, IResourceStack::amount,
+                factory::create);
```

Comment:
Could we make `IStackFactory` implement `BiFunction<R, Integer, S>` and then have that be a defaulted method it overrides to redirect it to the named `create` method? That way we can just pass the factory to this and the other codecs instead of having to do `factory::create`